openapi-diff/src/modeler/AutoRest.Swagger/Model/Schema.cs

```diff
@@ -184,27 +184,58 @@ Schema previous
         }
 
         /// <summary>
-        /// Comapares list of required properties of this model
+        /// Compares list of required properties of this (i.e. "new") model to the previous (i.e. "old") model.
+        ///
+        /// If the new model has new required properties that are not readOnly, it is a breaking change.
+        /// In case there are any issues with property definitions, like:
+        /// - properties being listed as required but not actually defined,
+        /// - or properties changing their read-only status.
+        /// Then we conservatively also treat such cases as breaking changes.
+        ///
+        /// Related:
+        /// https://stackoverflow.com/questions/40113049/how-to-specify-if-a-field-is-optional-or-required-in-openapi-swagger
+        /// https://github.com/Azure/azure-sdk-tools/issues/7184
         /// </summary>
-        /// <param name="context">Comaprision Context</param>
+        /// <param name="context">Comparison Context</param>
         /// <param name="priorSchema">Schema of the old model</param>
         private void CompareRequired(ComparisonContext<ServiceDefinition> context, Schema priorSchema)
         {
+            // If Required list is null then no properties are required in the new model,
+            // so there are no breaking changes.
             if (Required == null)
             {
                 return;
             }
 
-            if (Required != null && priorSchema.Required == null)
+            var newRequiredNonReadOnlyPropNames = new List<string>();
+            foreach (string requiredPropName in Required)
             {
-                context.LogBreakingChange(ComparisonMessages.AddedRequiredProperty, String.Join(", ", Required));
-                return;
+                Properties.TryGetValue(requiredPropName, out Schema propSchema);
+                priorSchema.Properties.TryGetValue(requiredPropName, out Schema priorPropSchema);
+                if (propSchema is null || priorPropSchema is null)
```

Comment:
@mikekistler due to this check:

`priorPropSchema is null`

this rule will flag on any properties that are `required` but not actually defined, even if there were no changes to them.

Let me know if you don't want that.

Reply:
It is legal (but very uncommon) to name properties in `required` that are not defined properties in the schema.

I think the desired behavior is to only flag properties when then changed from not-required to required, regardless of whether they are defined properties.

Reply:
@mikekistler I will adjust as follows:

1. If the property **was** required and **is** required: do not report a breaking change. This is regardless of property being actually defined or not.
2. If the property **was not** required and **is** required then:
  - If the property **was defined and read-only** and **is defined and read-only** then do not report breaking change.
  - In all other cases, report breaking changes.